src/classes/CW_AudienceGroupCompControlerTest.cls

```diff
@@ -0,0 +1,270 @@
+			
+			compCtrl.doNext();
+			system.assertEquals(2     ,compCtrl.getPageNumber());
+			system.assertEquals(true  ,compCtrl.getHasPrevious());
+			system.assertEquals(true  ,compCtrl.getHasNext());
+
+			
+			compCtrl.doPrevious();
+			system.assertEquals(false  ,compCtrl.getHasPrevious());
+			system.assertEquals(true   ,compCtrl.getHasNext());
+
+			compCtrl.editGroups();
+			system.assertEquals(false ,compCtrl.getOutputMode());
+			system.assertEquals(false ,compCtrl.savedByParent);
+			compCtrl.savedByParent =true;
+			system.assertEquals(true  ,compCtrl.getOutputMode());
+			compCtrl.Cancel();
+			system.assertEquals(true  ,compCtrl.getOutputMode());
+	    }
+	}
+	static testMethod void WraperClasssControllerTest() 
+	{
+        initCustomSettings();
+        
+        CW_Campaign__c camp = TestDataUtil.createCampaign();
+        insert camp;
+        
+        List<CW_Audience_Group__c> listGroups = TestDataUtil.createAudienceGroups(camp.Id,5);
+        insert listGroups;
+        
+        User thisUser = [SELECT Id FROM User WHERE Id = :UserInfo.getUserId()];
+    	
+    	Test.startTest();
+    	
+		system.runAs(thisUser)
+	    {  
+
+			List<sObject> RecList= new List<sObject>();
+			
+			//Initialization 
+	    	CW_AudienceGroupCompControler compCtrl= new CW_AudienceGroupCompControler();
+	    	CompCtrl.campId=camp.Id;
+			compCtrl.audienceGroups=RecList;
+			compCtrl.isReadOnly=false;
+			
+			CW_AudienceGroupCompControler.audienceGroupWrapper rec=compCtrl.campGroups[1];
+			
+			system.assertEquals(0     , rec.members);
+			system.assertEquals(false , rec.getIsUsed());
+			system.assertEquals(true , rec.isControlGroup);
+			rec.isControlGroup=false;
+			rec.aGroup.Control_Group__c=listGroups[0].Id;
+			system.assertEquals(false , rec.isControlGroup);
+			rec.isControlGroup=true;
+			system.assertEquals(true , rec.isControlGroup);
+			system.assertEquals(null , rec.aGroup.Control_Group__c);
+			rec.isControlGroup=false;
+	    	system.assertEquals(listGroups[0].Id , rec.aGroup.Control_Group__c);
+	    	
+	    }
+	}
+
+	static testMethod void editSaveGroupsTest() 
+    {
+        initCustomSettings();
+        
+        CW_Campaign__c camp = TestDataUtil.createCampaign();
+        camp.Recurrence_Type__c = 'Single';
+        insert camp;
+        
+        CW_Schedule__c sche = new CW_Schedule__c();
+        sche.Marketing_Campaign__c = camp.Id;
+		sche.Start_Date__c = camp.Campaign_Start_Date__c;
+		sche.End_Date__c = camp.Campaign_Start_Date__c;
+		sche.Evaluation_Start_Date__c = camp.Campaign_Start_Date__c;
+		sche.Evaluation_End_Date__c = camp.Campaign_Start_Date__c.addDays(60);
+		insert sche;
+        
+        List<CW_Audience_Group__c> listGroups = TestDataUtil.createAudienceGroups(camp.Id,2);
+        listGroups[0].Name = 'My Control Group';
+        listGroups[1].Name = 'My Target Group';
+        insert listGroups;
+        
+        User thisUser = [SELECT Id FROM User WHERE Id = :UserInfo.getUserId()];
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.